### ISSUE TITLE
[Unencrypted] Invited person does not see messages

### ISSUE THREAD

**Issue by MichalNemec** [element-hq/synapse#17155]
### Description

I need to see messages before accepting the conversation to prevent spam.
I also asked on matrix channel twice, no reply.

### Steps to reproduce

I used element app:

1. created a room
2. made the room public
3. "You made future room history visible to anyone."
4. "You changed the room name to PUBLIC"
5. "You changed the topic to "?"
6. "You invited [testing user]"
7. sent "hello" message
9. see empty screen, no messages.

### Homeserver

self-hosted

### Synapse Version

v1.105.0

### Installation Method

Docker (matrixdotorg/synapse)

### Database

Postgres

### Workers

Single process

### Platform

Ubuntu

### Configuration

-

### Relevant log output

```shell
-
```

### Anything else that would be useful to know?

I am using matrix-dart-sdk in flutter app, but even element app is not working properly on this usecase.

When someone messages me (unencrypted), then i need to see messages before accepting conversation, Because there is no way of knowing if its spam or not. Going the other way to check message etc before sending it is not efficient.
related issue on matrix-dart-sdk https://github.com/famedly/matrix-dart-sdk/issues/1756

**Comment by reivilibre** [comment]
I think Element intentionally doesn't want to show room history in invites to prevent abuse. If in doubt you'd need to ask them though.

If you are writing your own app, then for public, world-readable rooms that your server is already joined to, it should be possible to request history from the room even though you haven't joined it. (you might need to make the room guest-accessible, I'm not sure).

See this section of the spec: 'Room Previews' https://spec.matrix.org/v1.10/client-server-api/#room-previews

Otherwise as far as I know this is not currently possible; invited members don't receive any rights to see room history, just a small amount of room state. You'd need to search for or make a spec change proposal to change this (see https://github.com/matrix-org/matrix-spec-proposals/).


Unless I got the wrong end of the stick I think this answers your issue

**Comment by MichalNemec** [comment]
So to have exactly the same behavior like for example instagram, what someone messages me and i see the messages before accepting is impossible then, got it.

This example what i shown above is to just show that its not working properly from my pov.

**Comment by anoadragon453** [comment]
Correct, it's currently impossible to see room history in all cases before you've joined a room.

If I were trying to solve this problem, I would start at the [Matrix Spec](https://spec.matrix.org/latest/) side. I'd write a [Matrix Spec Change](https://github.com/matrix-org/matrix-spec-proposals/?tab=readme-ov-file#matrix-specification-proposals) to allow including some recent room history in the invite, much as we do with [stripped state events](https://spec.matrix.org/v1.12/client-server-api/#stripped-state). That is the mechanism that allows you to see a room's name/avatar/etc. in the invite.

Because of the nature of untrusted, decentralised systems and how Matrix works, it wouldn't be possible for your homeserver to verify that the included messages are actually real or really a part of the room. But I'd argue that it's not necessary for the homeserver to verify, as it's easy enough for the user to do once you've joined.

Clients could then require user interaction to show message contents, which would mitigate some abuse scenarios.

TL;DR this is a spec problem first, Synapse problem second. If you don't want to write the proposal yourself, file an issue on [matrix-org/matrix-spec](https://github.com/matrix-org/matrix-spec/issues) and someone else may come along and do so.
